### PR TITLE
Fix/amm bugs

### PR DESCRIPTION
**Pull request by LukasDeco**
- latest futarchy-ts
- validation for liquidity depositing(must validate base compared to quote or txn will fail)
- decimals bugs
- pipeline for npm publish

### CODE REVIEW

**Comment by LukasDeco** [comment]
@R-K-H there are issues encountered with withdrawal. I can't really debug them to well they seem to my eyes like maybe an issue within `futarchy-ts` although I could be wrong about that.

**Withdrawal error:**
`TypeError: src.toArrayLike is not a function`
image

Feels like something at the SDK level because as far as I can see the parameters are pretty simple and we are using the futarchy-ts SDK correctly.

File: lib/client/rpc/market-clients/ammMarkets.ts [hunk 1]
```diff
@@ -78,82 +74,106 @@ export class FutarchyAmmMarketsRPCClient implements FutarchyAmmMarketsClient {
...
     const minBaseAmount = 23;
     const minQuoteAmount = 20;
     const ix = this.ammClient.removeLiquidityIx(
-      ammAddr,
-      ammAcount.baseMint,
-      ammAcount.quoteMint,
+      ammMarket.publicKey,
+      ammMarket.baseMint,
+      ammMarket.quoteMint,
       lpTokensToBurn,
```
Comment (R-K-H) [inline lib/client/rpc/market-clients/ammMarkets.ts:177]: This can't be a number it needs to be a BN.

Comment (R-K-H) [reply lib/client/rpc/market-clients/ammMarkets.ts:177]: ammAddr: PublicKey,
    baseMint: PublicKey,
    quoteMint: PublicKey,
    lpTokensToBurn: BN,
    minBaseAmount: BN,
    minQuoteAmount: BN

File: lib/client/indexer/market-clients/ammMarkets.ts [hunk 2]
```diff
@@ -24,55 +26,72 @@ export class FutarchyIndexerAmmMarketsClient
...
     swapType: SwapType,
     inputAmount: number,
     outputAmountMin: number
   ): Promise<string[]> {
     return this.rpcMarketsClient.swap(
-      ammAddr,
+      ammMarket,
       swapType,
       inputAmount,
       outputAmountMin
```
Comment (R-K-H) [inline lib/client/indexer/market-clients/ammMarkets.ts:84]: Just as a reference, this needs to be calculated on the UI side OR / AND we can pass in a slippage param to the SDK such that we validate it in both areas... 

You output amount is the `estimated - (estimated * slippage)`

File: lib/client/indexer/market-clients/ammMarkets.ts [hunk 3]
```diff
@@ -24,55 +26,72 @@ export class FutarchyIndexerAmmMarketsClient
...
-    baseAmount?: number | undefined
-  ): Promise<string[]> {
-    return this.rpcMarketsClient.addLiquidity(ammAddr, quoteAmount, baseAmount);
+    ammMarket: AmmMarket,
+    quoteAmount: number,
+    baseAmount: number
+  ): Promise<string[] | LiquidityAddError> {
+    return this.rpcMarketsClient.addLiquidity(
+      ammMarket,
+      quoteAmount,
```
Comment (R-K-H) [inline lib/client/indexer/market-clients/ammMarkets.ts:49]: Just as a reference, this needs to be calculated on the UI side OR / AND we can pass in a slippage param to the SDK such that we validate it in both areas... 

You quoteAmount amount is the `quoteAmount - (quoteAmount * slippage)`

You quoteAmount amount is the `baseAmount - (baseAmount * slippage)`

File: lib/client/rpc/market-clients/ammMarkets.ts [hunk 4]
```diff
@@ -78,82 +74,106 @@ export class FutarchyAmmMarketsRPCClient implements FutarchyAmmMarketsClient {
...
-      ammAcount.quoteMint,
-      baseAmountCasted,
-      quoteAmountCasted,
-      minLpTokens,
+      ammMarket.publicKey,
+      ammMarket.baseMint,
+      ammMarket.quoteMint,
+      quoteAmountArg,
+      baseAmountArg,
+      lpTokensToMint,
```
Comment (R-K-H) [inline lib/client/rpc/market-clients/ammMarkets.ts:144]: This implementation is a misnomer, we need to use the actual values in here.

https://github.com/metaDAOproject/futarchy/blob/3b38e539f1fe50fef77c234961be8b3acaead7ce/futarchy-ts/src/AmmClient.ts#L237C5-L239C21

maxBaseAmount needs to be **multiplied** by  `(1 - slippage)`
minLpTokens needs to be  needs to be **multiplied** by `(1 - slippage)`

File: lib/client/rpc/market-clients/ammMarkets.ts (hunk 1 above)
Comment (R-K-H) [inline lib/client/rpc/market-clients/ammMarkets.ts:177]: So for `minBaseAmount` we need to take value **multiplied** by `(1 - slippage)`
For `minQuoteAmount` we need to take value **multiplied** by `(1 - slippage)`

File: lib/client/indexer/market-clients/ammMarkets.ts (hunk 2 above)
Comment (R-K-H) [reply lib/client/indexer/market-clients/ammMarkets.ts:84]: `outputAmountMin` that value should be the result of  what we have **multiplied** by `(1 - slippage)`

File: lib/client/indexer/market-clients/ammMarkets.ts (hunk 3 above)
Comment (R-K-H) [reply lib/client/indexer/market-clients/ammMarkets.ts:49]: These names should likely match what we have coming from the IX in the ts repo aas it's important to understanding the concepts.

File: lib/client/rpc/market-clients/ammMarkets.ts [hunk 5]
```diff
@@ -78,82 +74,106 @@ export class FutarchyAmmMarketsRPCClient implements FutarchyAmmMarketsClient {
...
+    const quoteAmountArg = new BN(
+      quoteAmount *
+        new BN(10).pow(new BN(ammMarket.quoteToken.decimals)).toNumber()
+    );
+    const baseAmountArg = new BN(
+      maxBasAmount *
+        new BN(10).pow(new BN(ammMarket.baseToken.decimals)).toNumber()
+    );
+    // base passed in should be ammBaseAmount honestly...
+    const baseReserve = new BN(ammMarket.baseAmount);
```
Comment (R-K-H) [inline lib/client/rpc/market-clients/ammMarkets.ts:91]: Not sure what we're tying to do here, just wanted to double check as the values of the quoteAmount and baseAmount aren't clear what that's expected to be.

File: lib/client/rpc/market-clients/ammMarkets.ts [hunk 6]
```diff
@@ -78,82 +74,106 @@ export class FutarchyAmmMarketsRPCClient implements FutarchyAmmMarketsClient {
     }
   }
 
+  validateAddLiquidity(
+    ammMarket: AmmMarket,
+    quoteAmount: number,
+    maxBasAmount: number
```
Comment (R-K-H) [inline lib/client/rpc/market-clients/ammMarkets.ts:80]: `maxBaseAmount` is likely what you wanted to pass as the name.

File: lib/client/rpc/market-clients/ammMarkets.ts [hunk 7]
```diff
@@ -78,82 +74,106 @@ export class FutarchyAmmMarketsRPCClient implements FutarchyAmmMarketsClient {
...
+  }
+
   async addLiquidity(
-    ammAddr: PublicKey,
-    quoteAmount?: number,
-    baseAmount?: number
-  ) {
+    ammMarket: AmmMarket,
+    quoteAmount: number,
+    maxBasAmount: number
```
Comment (R-K-H) [inline lib/client/rpc/market-clients/ammMarkets.ts:110]: `maxBaseAmount` here too..

File: lib/client/indexer/market-clients/ammMarkets.ts (hunk 2 above)
Comment (LukasDeco) [reply lib/client/indexer/market-clients/ammMarkets.ts:84]: This has been fixed now in here: https://github.com/metaDAOproject/metadao-frontend/pull/56/commits/22de20d4c41fffc159fad43796d805aeacb8232e

File: lib/client/rpc/market-clients/ammMarkets.ts (hunk 6 above)
Comment (LukasDeco) [reply lib/client/rpc/market-clients/ammMarkets.ts]: Whoopsie fixed this

File: lib/client/rpc/market-clients/ammMarkets.ts (hunk 7 above)
Comment (LukasDeco) [reply lib/client/rpc/market-clients/ammMarkets.ts]: whoops fixed

File: lib/client/rpc/market-clients/ammMarkets.ts (hunk 1 above)
Comment (LukasDeco) [reply lib/client/rpc/market-clients/ammMarkets.ts:178]: done now

File: lib/client/indexer/market-clients/ammMarkets.ts (hunk 3 above)
Comment (LukasDeco) [reply lib/client/indexer/market-clients/ammMarkets.ts:49]: Names have been changed to match the amm client. LMK if you see something wrong here still.

File: lib/client/rpc/market-clients/ammMarkets.ts (hunk 4 above)
Comment (LukasDeco) [reply lib/client/rpc/market-clients/ammMarkets.ts]: Names have been changed, but I will do slippage math on frontend

File: lib/client/rpc/market-clients/ammMarkets.ts (hunk 1 above)
Comment (LukasDeco) [reply lib/client/rpc/market-clients/ammMarkets.ts:178]: will do math on FE or... I might have some questions on where these values come from

File: lib/client/rpc/market-clients/ammMarkets.ts (hunk 4 above)
Comment (LukasDeco) [reply lib/client/rpc/market-clients/ammMarkets.ts]: psyche, I will do slippage math in the SDK

File: lib/client/rpc/market-clients/ammMarkets.ts [hunk 8]
```diff
@@ -78,85 +75,146 @@ export class FutarchyAmmMarketsRPCClient implements FutarchyAmmMarketsClient {
...
-    const quoteAmountCasted = new BN(quoteAmount).mul(
-      new BN(10).pow(new BN(quoteDecimals))
+    const maxBaseAmountArg = new BN(
+      maxBaseAmount *
+        new BN(10).pow(new BN(ammMarket.baseToken.decimals)).toNumber()
     );
 
-    const baseInfo = await this.rpcProvider.connection.getAccountInfo(
-      ammAcount.baseMint
+    const quoteAmountWithSlippage = calculateMaxWithSlippage(
```
Comment (R-K-H) [inline lib/client/rpc/market-clients/ammMarkets.ts:141]: So this isn't correct. Quote amount doesn't have slippage, just a future todo to fix.

Comment (R-K-H) [reply lib/client/rpc/market-clients/ammMarkets.ts:141]: maxBaseAmount: BN,
minLpTokens: BN,

Are the two that require slippage calcualtion.

**Review by R-K-H — APPROVED** [review]
Have one issue just to put a further TODO to solve, but otherwise LGTM

File: lib/client/rpc/market-clients/ammMarkets.ts (hunk 8 above)
Comment (R-K-H) [reply lib/client/rpc/market-clients/ammMarkets.ts:141]: I believe this is the issue https://github.com/metaDAOproject/metadao-frontend/issues/60

**Pull request merged**